### PR TITLE
add _draw_auto_regression_np

### LINKED ISSUE #150: auto regression: add numpy-based 'predict' function
Extract the code to 'predict' auto regression and write a numpy-based function. This is used in two places:

- https://github.com/MESMER-group/mesmer/blob/9def7422e2b0b77a105a6c1443e3d7c1eb6b0858/mesmer/create_emulations/create_emus_lv.py#L112
- https://github.com/MESMER-group/mesmer/blob/9def7422e2b0b77a105a6c1443e3d7c1eb6b0858/mesmer/create_emulations/create_emus_gv.py#L111

The two approaches are slightly different - one is based on a multivariate gaussian and the other on a univariate. However, it should be possible to unify them. My idea would be to do both using `multivariate_normal` and then `squeeze` the 1D case.

```python
import numpy as np

np.random.seed(1)
a = np.random.normal(loc=0, scale=1, size=3)

np.random.seed(1)
b = np.random.multivariate_normal([0], [[1]], size=3)

np.testing.assert_allclose(a, b.squeeze())
```

### PR DESCRIPTION
**Pull request by mathause**
- [x] Closes #150
 - [x] Tests added
 - [x] Passes `isort . && black . && flake8`
 - [x] Fully documented, including `CHANGELOG.rst`

---

A first stab at factoring out 'predicting' an AR process. It does not look very clean to me at the moment but it works.

Comments
- The coefs need to be 2D with shape `n_coefs x n_gridpoints`
- The `covariance` must be passed and not the `standard_deviation`
- `emus_lv_tmp` should be renamed
- `emus_lv_tmp` has shape `n_emus x (n_ts + buffer) x n_cells` - maybe it would be better if this was `n_emus x n_cells x (n_ts + buffer)` (so we can use `@` to multiply and sum, this would require to transpose the coefs)

### CODE REVIEW

File: mesmer/create_emulations/create_emus_lv.py [hunk 1]
```diff
@@ -180,39 +182,29 @@ def create_emus_lv_AR1_sci(emus_lv, params_lv, preds_lv, cfg):
...
-                    + params_lv["AR1_coef"][targ] * emus_lv_tmp[:, t - 1, :]
-                    + innovs[:, t, :]
-                )
-            emus_lv_tmp = emus_lv_tmp[:, buffer:, :]
-
-            print("Create the full local variability emulations")
-            emus_lv[scen][targ] += emus_lv_tmp
 
+            emus_lv[scen][targ] += emus_ar.squeeze()
+            print(f"{emus_lv[scen][targ]=}")
```
Comment (mathause) [inline mesmer/create_emulations/create_emus_lv.py:207]:
```suggestion
```

**Comment by mathause** [comment]
@yquilcaille - the PR is not finished yet but if you want to take a look at it, that'd be welcome.

File: mesmer/create_emulations/create_emus_lv.py [hunk 2]
```diff
@@ -180,39 +182,28 @@ def create_emus_lv_AR1_sci(emus_lv, params_lv, preds_lv, cfg):
...
-                size=[nr_emus_v, nr_ts_emus_stoch_v + buffer],
-            )
-
-            print(
-                "Compute the contribution to emus_lv by the AR(1) process with the "
-                "spatially correlated innovations"
+            emus_ar = _predict_auto_regression_np(
+                intercept=intercept,
+                coefs=coefs,
+                covariance=covariance,
```
Comment (mathause) [inline mesmer/create_emulations/create_emus_lv.py:198]:
```suggestion
                intercept=params_lv["AR1_int"][targ],
                # reshape to n_coefs x n_cells
                coefs=params_lv["AR1_coef"][targ][np.newaxis, :],
                covariance=params_lv["loc_ecov_AR1_innovs"][targ],
```

File: mesmer/create_emulations/create_emus_lv.py [hunk 3]
```diff
@@ -180,39 +182,28 @@ def create_emus_lv_AR1_sci(emus_lv, params_lv, preds_lv, cfg):
...
             if len(emus_lv[scen]) == 0:
-
                 emus_lv[scen][targ] = np.zeros(nr_emus_v, nr_ts_emus_stoch_v, nr_gps)
 
-            # ensure reproducibility
-            np.random.seed(seed)
+            intercept = params_lv["AR1_int"][targ]
+            # reshape to n_coefs x n_cells
+            coefs = params_lv["AR1_coef"][targ][np.newaxis, :]
+            covariance = params_lv["loc_ecov_AR1_innovs"][targ]
```
Comment (mathause) [inline mesmer/create_emulations/create_emus_lv.py:190]:
```suggestion
```

File: mesmer/create_emulations/create_emus_gv.py [hunk 4]
```diff
@@ -157,33 +159,24 @@ def create_emus_gv_AR(params_gv, nr_emus_v, nr_ts_emus_v, seed):
...
-            + innovs_emus_gv[:, t]
-        )
-
-    # remove buffer
-    emus_gv = emus_gv[:, buffer:]
-
-    return emus_gv
+    emus_gv = _predict_auto_regression_np(
+        intercept=ar_int,
+        coefs=ar_coefs[:, np.newaxis],
```
Comment (mathause) [inline mesmer/create_emulations/create_emus_gv.py:173]:
```suggestion
        # reshape to n_coefs x n_cells
        coefs=ar_coefs[:, np.newaxis],
```

**Comment by mathause** [comment]
Marking this ready for review. @znicholls do you want to have a look?

File: mesmer/core/auto_regression.py [hunk 5]
```diff
@@ -2,6 +2,72 @@
...
+        Drawn realizations of the specified autoregressive process. The array has shape
+        n_samples x n_ts x n_cells.
+
+    Notes
+    -----
+    As this is not an deterministic function it is not called `predict`. "Predicting"
+    an autoregressive process does not include the innovations and therefore asymptotes
+    towards certain value (in constrast to this function).
+    """
+
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:42]:
```suggestion
```

File: mesmer/core/auto_regression.py [hunk 6]
```diff
@@ -2,6 +2,72 @@
...
+    intercept = np.array(intercept)
+    covariance = np.atleast_2d(covariance)
+
+    # coeffs must be ar_order x n_cells
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:46]:
```suggestion
    # coeffs assumed to be ar_order x n_cells
```

File: mesmer/core/auto_regression.py [hunk 7]
```diff
@@ -2,6 +2,72 @@
...
+    ar_order, n_cells = coefs.shape
+
+    # TODO: allow arbitrary lags? (e.g., [1, 12]) -> need to pass `ar_lags`
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:49]: Make an issue and reference in comment?

File: mesmer/core/auto_regression.py [hunk 8]
```diff
@@ -2,6 +2,72 @@
...
+    ar_lags = np.arange(1, ar_order + 1, dtype=int)
+
+    # ensure reproducibility
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:52]:
```suggestion
    # ensure reproducibility (TODO: clarify approach to this see #35)
```

File: mesmer/core/auto_regression.py [hunk 9]
```diff
@@ -2,6 +2,72 @@
 import xarray as xr
 
 
+def _draw_auto_regression_np(
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:5]: Can we change the name given that this is not really an auto-regression draw (it's auto-regression plus the spatial covariance)? Perhaps `_draw_auto_regression_with_innovations_np`?

File: mesmer/core/auto_regression.py [hunk 10]
```diff
@@ -2,6 +2,72 @@
...
+    *, intercept, coefs, covariance, n_samples, n_ts, seed, buffer
+):
+    """draw time series of an autoregressive process
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:8]:
```suggestion
    """Draw time series of an autoregressive process plus innovations
```

Can we add a one line definition of 'innovations' in here (or a reference to somewhere which explains this term, I've never seen it before)

File: mesmer/core/auto_regression.py [hunk 11]
```diff
@@ -2,6 +2,72 @@
...
+    np.random.seed(seed)
+
+    innovations = np.random.multivariate_normal(
+        mean=np.zeros(n_cells),
+        cov=covariance,
+        size=[n_samples, n_ts + buffer],
+    )
+
+    out = np.zeros([n_samples, n_ts + buffer, n_cells])
+    for t in range(ar_order + 1, n_ts + buffer):
```
Comment (znicholls) [inline mesmer/core/auto_regression.py:62]: Is there no way to vectorise this (maybe not, just seems odd for us to have loops)? Can also be done in a follow up PR

**Review by znicholls — APPROVED** [review]
What a beauty! A few comments but happy for you to make the call about whether you just merge this and create issues for anything that is tricky to address or just address things now.

**Review by mathause — COMMENTED** [review]
Thanks for the fast review!

---

The 'innovations' in a auto regression is the error or noise term. I think the idea behind the name is that for each time step this is the 'new' element i.e. the _innovation_ - i.e. in 

`$x_t = x_{t - 1} * \phi + \epsilon_t$`

it would be the `\epsilon_t$`.

I'll add something to the docstring.

---

I have unified the the global (= 1 cell without correlation, obviously) and local (= n cells with auto correlation) case. So it's only "possibly" correlated. I am not sure if it is better to have one or two functions.

```python
class AutoRegression:

    def draw(...): ...

# OR

    def draw(...): ...

    def draw_correlated(...): ...

# OR

    def draw_1d(...): ...

    def draw_2d(...): ...

# OR ???

```

**Comment by znicholls** [comment]
> I am not sure if it is better to have one or two functions

I would go one and just add something like the following to the description of the `coeff` parameter: "If `coeff` is set to zero then this is just a standard sample from an autoregressive process." That will help users know that this function can be used without innovations if they want. If we get more complicated, we might want to refactor but for now it should be fine.

File: mesmer/core/auto_regression.py [hunk 12]
```diff
@@ -2,6 +2,72 @@
 import xarray as xr
 
 
+def _draw_auto_regression_np(
+    *, intercept, coefs, covariance, n_samples, n_ts, seed, buffer
+):
+    """draw time series of an autoregressive process
+
+    Parameters
+    ----------
```
[the remark this answers is not in the document]
Comment (znicholls) [reply mesmer/core/auto_regression.py:13]: Add a docstring of what expected args are (there's a `*` in the function signature, but the docstring doesn't describe what is expected or where these args go). So the docstring would look like something like

```
Parameters
---------------
args
    Arguments passed through to something
```

Looking at it more now, the `*` in the function signature isn't used so could be redundant?

[the remark this answers is not in the document]
Comment (mathause) [reply mesmer/core/auto_regression.py:18]: This indicates that all arguments are keyword only.

**Review by mathause — COMMENTED** [review]
I renamed the function, added an explanatory sentence and tried to clear up the names. `n_cells` is now called `n_coefs` - I hope this is not confused with the order of the AR process.

**Comment by znicholls** [comment]
Very nice. Other option would be `n_gridpoints`. I'm happy with whatever so once you choose what you like best, merge away

**Pull request merged**